lib/truffle/socket/socket.rb

```diff
@@ -341,6 +341,10 @@ def connect_nonblock(sockaddr, exception: true)
 
     status = Truffle::Socket::Foreign.connect(descriptor, sockaddr)
 
+    if Errno.errno == Errno::EISCONN::Errno
+      raise Errno::EISCONN
+    end
```

Comment:
`errno` should only be read in the error case, otherwise it could have a value from the previous syscall.

`Truffle::Socket::Error.write_nonblock('connect(2)')` below handles all errnos, but the logic for `exception: false` is buggy, as I think `:wait_writable` should only be returned for EAGAIN and EWOULDBLOCK.

How about:
```ruby
if status < 0
  if !exception and Truffle::POSIX::TRY_AGAIN_ERRNOS.include?(Errno.errno)
    :wait_writable
  else
    Truffle::Socket::Error.write_nonblock('connect(2)')
  end
```

Reply:
`Truffle::Socket::Error.write_nonblock` does the magic and wraps the error with `IO::WaitWritable` if it's EAGAIN, EWOULDBLOCK or EINPROGRESS, but otherwise let the exception go through, so I think it's correct.
https://github.com/oracle/truffleruby/blob/78231bc55170f37dd5394f87b997cfcabbcfd7b7/lib/truffle/socket/truffle/error.rb#L52-L54
Anyway, the spec will show it if it's not.

Reply:
You specifically need to raise `Errno::EISCONN` even if `exception: false`.

Reply:
Yes, the `Errno.handle` in `write_nonblock` does just that, and that's how the spec mentioned below passes with `exception: false`.

Reply:
Actually, it's more complicated than that, MRI returns 0 for `@client.connect_nonblock(@server.getsockname, exception: false)` is `@client` is already connected, due to special handling for EISCONN:
https://github.com/ruby/ruby/blob/v2_4_5/ext/socket/socket.c#L442-L471

Reply:
Interesting, maybe it was an older version I had to work around. Definitely prefer no exception path.